### PR TITLE
LPD-35811 Create a filter to check CSP implementation at CI

### PR DESCRIPTION
**Pull request by izaera**
**On hold**: we cannot forward this until all CSP stuff is merged (including the SF rule).

----
Testing how it goes...

This PR should fail a lot of tests since style, link and attributes are not yet processed by CSP.

### CODE REVIEW

**Comment by izaera** [comment]
ci:report:5618414

**Comment by izaera** [comment]
ci:test:relevant

**Comment by izaera** [comment]
ci:test:report 16942751

**Comment by izaera** [comment]
ci:report:16942751

**Comment by izaera** [comment]
This  is more or less ready for review. In the sense that it is finished and only needs polishing the logic for the filter to be activated.

**Comment by izaera** [comment]
ci:test:stable

**Comment by izaera** [comment]
Let's see what happens when tests fail :point_down:

**Comment by izaera** [comment]
I've removed the unconditional activation of the filter. Let's see if using `PortalRunMode.isTestMode()` does the trick at CI...

**Comment by izaera** [comment]
ci:test:stable

**Comment by izaera** [comment]
Here's how a failed poshi test will look like when a CSP violation is found:

image

**Comment by izaera** [comment]
And here's how a failed playwright test would look like:

image

**Comment by herrardo** [comment]
> Here's how a failed poshi test will look like when a CSP violation is found:
> 
> image

In this Poshi test could we say why an 'onload' attribute is considered an error. Somebody out from CSP initiative might not know why we are preventing this practices, something like:

"Please check your changes to the source code and make sure do not use the onload attribute in any Java, JSP, or template. This will allow to enable CSP with restictive values for diverse directives"

What do you think?

**Comment by izaera** [comment]
:warning: Agreed with Brian to turn the filter on for `localhost` too (for now) so that CSP violations are detected on CI and local development.

Need to add it to the code...

**Comment by izaera** [comment]
ci:test:sf

**Comment by izaera** [comment]
ci:test:stable

**Comment by izaera** [comment]
> Stable should fail if PortalRunMode.isTestMode() is running correctly...

It did. So good :tada:

**Comment by izaera** [comment]
This PR is OK to forward but waiting for https://github.com/brianchandotcom/liferay-portal/pull/154709

**Comment by izaera** [comment]
ci:test:sf

**Comment by izaera** [comment]
ci:test:relevant

**Comment by izaera** [comment]
This is how a CSP violation looks while developing locally:

image

File: modules/apps/portal-security/portal-security-content-security-policy/src/main/java/com/liferay/portal/security/content/security/policy/internal/servlet/filter/ContentSecurityPolicyHttpServletResponse.java [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+import java.io.OutputStreamWriter;
+import java.io.PrintWriter;
+
+import javax.servlet.ServletOutputStream;
+import javax.servlet.WriteListener;
+import javax.servlet.http.HttpServletResponse;
+import javax.servlet.http.HttpServletResponseWrapper;
+
+/**
+ * @author Olivér Kecskeméty
```
Comment (bryceosterhaus) [inline modules/apps/portal-security/portal-security-content-security-policy/src/main/java/com/liferay/portal/security/content/security/policy/internal/servlet/filter/ContentSecurityPolicyHttpServletResponse.java:19]: who dis

File: modules/apps/portal-security/portal-security-content-security-policy/src/main/java/com/liferay/portal/security/content/security/policy/internal/servlet/filter/ContentSecurityPolicyTestFilter.java [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+public class ContentSecurityPolicyTestFilter extends BasePortalFilter {
+
+	@Override
+	public boolean isFilterEnabled(
+		HttpServletRequest httpServletRequest,
+		HttpServletResponse httpServletResponse) {
+
+		if (PortalRunMode.isTestMode() ||
+			StringUtil.equals(
+				httpServletRequest.getServerName(), "localhost")) {
```
Comment (bryceosterhaus) [inline modules/apps/portal-security/portal-security-content-security-policy/src/main/java/com/liferay/portal/security/content/security/policy/internal/servlet/filter/ContentSecurityPolicyTestFilter.java:64]: So would this be on whenever developing locally? Would this also be turned on for customers?

Also, should this only be on if CSP is enabled? Or maybe I missed that some place else

File: modules/apps/portal-security/portal-security-content-security-policy/src/main/java/com/liferay/portal/security/content/security/policy/internal/servlet/filter/ContentSecurityPolicyTestFilter.java [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+				ContentSecurityPolicyNonceProviderUtil.getNonceAttribute(
+					httpServletRequest));
+			printWriter.write(">pre{margin:4em auto; max-width:1024px;");
+			printWriter.write("overflow:hidden;text-wrap:auto;} pre div{");
+			printWriter.write("background:lightgray;display:inline-block;");
+			printWriter.write("margin-bottom:2em;padding:.2em 1em;}</style>");
+			printWriter.write("<pre><center><div>Content ");
+			printWriter.write("Security Policy Check Failure</div></center>");
+			printWriter.write(HtmlUtil.escape(exception.getMessage()));
+			printWriter.write("</pre></body></html>");
```
Comment (bryceosterhaus) [inline modules/apps/portal-security/portal-security-content-security-policy/src/main/java/com/liferay/portal/security/content/security/policy/internal/servlet/filter/ContentSecurityPolicyTestFilter.java:136]: no multi-line templates in java?

File: modules/apps/portal-security/portal-security-content-security-policy/src/main/java/com/liferay/portal/security/content/security/policy/internal/servlet/filter/ContentSecurityPolicyHttpServletResponse.java (hunk 1 above)
Comment (izaera) [reply modules/apps/portal-security/portal-security-content-security-policy/src/main/java/com/liferay/portal/security/content/security/policy/internal/servlet/filter/ContentSecurityPolicyHttpServletResponse.java:19]: :shrug: 

The one that created that class. I'm moving it from another place. It already existed.

**Comment by izaera** [comment]
Since the LPD was discarded, I'm closing this PR.

**Pull request closed without merging**